### PR TITLE
GH1540: Update to latest Roslyn

### PR DESCRIPTION
**Pull request by patriksvensson**
This PR will do the following:

- [x] Update all projects to target `net462` instead of `net45`
- [x] Update Roslyn to the latest version
- [x] Remove old Roslyn (regular and experimental)
- [x] Remove the Mono scripting engine
- [x] Unify script engine between full framework and .net core
- [x] Make it build on CI again
- [x] Add loading assembly from stream for debugging purposes (see AssemblyLoader).
- [x] Verify that debugging works on full framework and .NET Core
- [x] Clean up the code a bit

### CODE REVIEW

File: build/paths.cake [hunk 1]
```diff
@@ -26,7 +26,7 @@ public class BuildPaths
         var buildDir = context.Directory("./src/Cake/bin") + context.Directory(configuration);
         var artifactsDir = (DirectoryPath)(context.Directory("./artifacts") + context.Directory("v" + semVersion));
         var artifactsBinDir = artifactsDir.Combine("bin");
-        var artifactsBinNet45 = artifactsBinDir.Combine("net45");
+        var artifactsBinNet45 = artifactsBinDir.Combine("net462");
```
Comment (mholo65) [inline build/paths.cake]: Also change variable name to `artifactsBinNet462`

**Comment by devlead** [comment]
@cake-build/cake-team I've removed some unused variables in path.cake and added so chocolatey/desktop CLR nuget content isn't hard coded any more. so we'll need to test so those still include all's that needed.

**Comment by devlead** [comment]
@cake-build/cake-team I've sorted the merge conflicts and rebased against latest develop.

**Comment by gep13** [comment]
Relates to #1540

**Comment by patriksvensson** [comment]
@mholo65 Did you have a fix for the debugging that we could use?

**Comment by mholo65** [comment]
@patriksvensson only the `AppDomain.AssemblyResolve` fix, but that feels wrong. IMO we need to investigate further and ask some help from Roslyn team. How is it working with `CoreCLR` debugging, same thing there?

**Comment by patriksvensson** [comment]
@mholo65 Not sure, and I won't have time to do anything with this PR until the weekend soonest.

**Comment by mholo65** [comment]
@patriksvensson Ok! I'll see what I can do!

**Comment by mholo65** [comment]
Seems to hit the `FileNotFoundException` when referencing addins and debugging. Will investigate further, trying to create a minimal repro and then contact Roslyn team if I can't figure it out myself.

Also, I seem to hit same exception when debugging with `Cake.CoreCLR` version 0.21.1. Could anyone of you verify this?

**Comment by filipw** [comment]
you might consider going straight for Roslyn 3 (nightly) and avoid hand-rolled debugging code altogether 🙂 https://github.com/dotnet/roslyn/pull/16489

**Comment by filipw** [comment]
actually, I forgot - that made it into Roslyn 2.3.0 already

**Comment by mholo65** [comment]
@filipw thanks for the heads up! @patriksvensson we should update to 2.3.x and check that out!

**Comment by mholo65** [comment]
@filipw got any examples for using the API? Will have a look at it tonight!

**Comment by gep13** [comment]
>@mholo65 said...
Could anyone of you verify this?

As discussed on slack, i have been able to verify this is happening.

**Comment by filipw** [comment]
@mholo65 try here https://github.com/filipw/dotnet-script/blob/master/src/Dotnet.Script.Core/ScriptCompiler.cs#L53

File: src/Cake/Scripting/Roslyn/RoslynScriptEngine.cs [hunk 2]
```diff
@@ -2,56 +2,33 @@
...
+        public RoslynScriptEngine(CakeOptions options, IAssemblyLoader loader, ICakeLog log)
         {
             _options = options;
-            _nightlyFactory = nightlyFactory;
-            _stableFactory = stableFactory;
+            _loader = loader;
             _log = log;
         }
 
         public IScriptSession CreateSession(IScriptHost host, IDictionary<string, string> arguments)
```
Comment (mholo65) [inline src/Cake/Scripting/Roslyn/RoslynScriptEngine.cs]: While we are at it breaking stuff. Maybe we should remove `arguments` from the interface?

Comment (patriksvensson) [inline src/Cake/Scripting/Roslyn/RoslynScriptEngine.cs]: Agree! 👍

**Comment by mholo65** [comment]
Roslyn now updated and debugging works. Unified the sessions also.

Thanks again @filipw for pointing us in the right direction!

**Comment by filipw** [comment]
great! 🎉

**Comment by patriksvensson** [comment]
@cake-build/cake-team OK! I think this PR is ready to be merged. Do you mind taking a look at it and also take it for a final spin if you have time.

File: src/Cake/Arguments/ArgumentParser.cs [hunk 3]
```diff
@@ -166,16 +166,6 @@ private bool ParseOption(string name, string value, CakeOptions options)
                 options.PerformDebug = ParseBooleanValue(value);
             }
 
-            if (name.Equals("mono", StringComparison.OrdinalIgnoreCase))
```
Comment (gep13) [inline src/Cake/Arguments/ArgumentParser.cs]: @patriksvensson @mholo65 I am not sure if these should be removed completely.  I think we should deprecate these commands. i.e. leave them in just now, but have them only emit an error that the Mono scripting engine has been removed with an explanation as to why.  Then in a future version, we can remove them completely.  Thoughts?

**Comment by gep13** [comment]
@patriksvensson @mholo65 related to the above inline comment, I did a quick search for the word `Mono` in the codebase, and looks like it still appears quite a few times.  Would it be worth going through these to decide what is, and isn't required?

File: src/Cake/Arguments/ArgumentParser.cs (hunk 3 above)
Comment (patriksvensson) [inline src/Cake/Arguments/ArgumentParser.cs]: Removing them won't break anything afaik.

**Comment by patriksvensson** [comment]
@cake-build/cake-team I feel that this PR is done now unless someone have objections or opinions about it.

🙌 🎉 🍰 🎈

**Comment by Meir017** [comment]
the class `CakeOptions` still has a `Mono` property ([here](https://github.com/patriksvensson/cake/blob/1988daabe1bfd440527c077cc777d4d601fbfbe6/src/Cake/CakeOptions.cs#L91)) and some tests ([here](https://github.com/patriksvensson/cake/blob/1988daabe1bfd440527c077cc777d4d601fbfbe6/src/Cake.Tests/Unit/Arguments/ArgumentParserTests.cs#L329))

**Comment by patriksvensson** [comment]
@Meir017 Yes, it's still used here: https://github.com/cake-build/cake/pull/1645/files#diff-3c3b3d790a1cf0ee07879f36796b4230R33

**Comment by devlead** [comment]
@mholo65 triggered an integration test on Ubuntu with Bitrise now.

**Comment by patriksvensson** [comment]
@gep13 Not sure why, but your review doesn't seem to close although the changes has been addressed.

**Comment by gep13** [comment]
@patriksvensson fixed

**Comment by patriksvensson** [comment]
@gep13 Thanks!
@devlead What was the result of the integration test?

**Comment by devlead** [comment]
@patriksvensson it didn't complete but don't think it's a PR issue will try to get them to rerun

File: build.cake [hunk 4]
```diff
@@ -144,13 +154,13 @@ Task("Copy-Files")
...
+    // .NET 4.6.2
     DotNetCorePublish("./src/Cake", new DotNetCorePublishSettings
     {
-        Framework = "net45",
+        Framework = "net462",
         VersionSuffix = parameters.Version.DotNetAsterix,
         Configuration = parameters.Configuration,
-        OutputDirectory = parameters.Paths.Directories.ArtifactsBinNet45,
+        OutputDirectory = parameters.Paths.Directories.ArtifactsBinFullFx,
         Verbosity = DotNetCoreVerbosity.Minimal
```
Comment (devlead) [inline build.cake]: @patriksvensson Currently getting an error with copy files on Linux
```
/usr/share/dotnet/sdk/1.0.4/Microsoft.Common.CurrentVersion.targets(1111,5): error MSB3644: The reference assemblies for framework ".NETFramework,Version=v4.6.2" were not found. To resolve this, install the SDK or Targeting Pack for this framework version or retarget your application to a version of the framework for which you have the SDK or Targeting Pack installed. Note that assemblies will be resolved from the Global Assembly Cache (GAC) and will be used in place of reference assemblies. Therefore your assembly may not be correctly targeted for the framework you intend. [/cake/src/Cake/Cake.csproj]
An error occurred when executing task 'Copy-Files'.
```

Adjusting copy files task like below seems to sort it
```diff
    // .NET 4.6.2
    DotNetCorePublish("./src/Cake", new DotNetCorePublishSettings
    {
        Framework = "net462",
        VersionSuffix = parameters.Version.DotNetAsterix,
        Configuration = parameters.Configuration,
        OutputDirectory = parameters.Paths.Directories.ArtifactsBinFullFx,
-        Verbosity = DotNetCoreVerbosity.Minimal,
+       MSBuildSettings = msBuildSettings
    });

    // .NET Core
    DotNetCorePublish("./src/Cake", new DotNetCorePublishSettings
    {
        Framework = "netcoreapp1.0",
        Configuration = parameters.Configuration,
        OutputDirectory = parameters.Paths.Directories.ArtifactsBinNetCore,
-        Verbosity = DotNetCoreVerbosity.Minimal,
+       MSBuildSettings = msBuildSettings
    });
```

**Comment by mholo65** [comment]
Squash is Ok as long as individual authors changes are kept intact. I.e. Should be squashed to 4 or 5 commits.

**Comment by devlead** [comment]
> Squash is Ok as long as individual authors changes are kept intact. I.e. Should be squashed to 4 or 5 commits.

@mholo65 agreed 👍

**Pull request merged**